File: facebook/src/com/facebook/login/LoginFragment.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.facebook.R;

/**
 * This Fragment is a necessary part of the overall Facebook login process
 * but is not meant to be used directly.
 *
 * @see com.facebook.FacebookActivity
 */

public class LoginFragment extends Fragment {
    static final String RESULT_KEY = "com.facebook.LoginFragment:Result";

    private static final String TAG = "LoginFragment";
    private static final String NULL_CALLING_PKG_ERROR_MSG =
            "Cannot call LoginFragment with a null calling package. " +
                    "This can occur if the launchMode of the caller is singleInstance.";
    private static final String EXTRA_REQUEST = "request";
    private static final String SAVED_LOGIN_CLIENT = "loginClient";

    private String callingPackage;
    private LoginClient loginClient;
    private LoginClient.Request request;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            loginClient = savedInstanceState.getParcelable(SAVED_LOGIN_CLIENT);
            loginClient.setFragment(this);
        } else {
            loginClient = new LoginClient(this);
        }

        loginClient.setOnCompletedListener(new LoginClient.OnCompletedListener() {
            @Override
            public void onCompleted(LoginClient.Result outcome) {
                onLoginClientCompleted(outcome);
            }
        });

        Activity activity = getActivity();
        if (activity == null) {
            return;
        }

        initializeCallingPackage(activity);
        if (activity.getIntent() != null) {
            request = (LoginClient.Request)
                    activity.getIntent().getParcelableExtra(EXTRA_REQUEST);
        }
    }

    @Override
    public void onDestroy() {
        loginClient.cancelCurrentHandler();
        super.onDestroy();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.com_facebook_login_fragment, container, false);

        loginClient.setBackgroundProcessingListener(
                new LoginClient.BackgroundProcessingListener() {
            @Override
            public void onBackgroundProcessingStarted() {
                view.findViewById(
                        R.id.com_facebook_login_activity_progress_bar).setVisibility(View.VISIBLE);
            }

            @Override
            public void onBackgroundProcessingStopped() {
                view.findViewById(
                        R.id.com_facebook_login_activity_progress_bar).setVisibility(View.GONE);
            }
        });

        return view;
    }

    private void onLoginClientCompleted(LoginClient.Result outcome) {
        request = null;

        int resultCode = (outcome.code == LoginClient.Result.Code.CANCEL) ?
                Activity.RESULT_CANCELED : Activity.RESULT_OK;

        Bundle bundle = new Bundle();
        bundle.putParcelable(RESULT_KEY, outcome);

        Intent resultIntent = new Intent();
        resultIntent.putExtras(bundle);

        // The activity might be detached we will send a cancel result in onDetach
        if (isAdded()) {
            getActivity().setResult(resultCode, resultIntent);
            getActivity().finish();
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        // If the calling package is null, this generally means that the callee was started
        // with a launchMode of singleInstance. Unfortunately, Android does not allow a result
        // to be set when the callee is a singleInstance, so we log an error and return.
        if (callingPackage == null) {
            Log.e(TAG, NULL_CALLING_PKG_ERROR_MSG);
            getActivity().finish();
            return;
        }

        loginClient.startOrContinueAuth(request);
    }

    @Override
    public void onPause() {
        super.onPause();

        getActivity().findViewById(R.id.com_facebook_login_activity_progress_bar).setVisibility(
                View.GONE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        loginClient.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putParcelable(SAVED_LOGIN_CLIENT, loginClient);
    }

    private void initializeCallingPackage(final Activity activity) {
        ComponentName componentName = activity.getCallingActivity();
        if (componentName == null) {
            return;
        }
        callingPackage = componentName.getPackageName();
    }

    static Bundle populateIntentExtras(LoginClient.Request request) {
        Bundle extras = new Bundle();
        extras.putParcelable(EXTRA_REQUEST, request);
        return extras;
    }
}

File: facebook-common/src/main/java/com/facebook/login/LoginFragment.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.facebook.common.R

/**
 * This Fragment is a necessary part of the overall Facebook login process but is not meant to be
 * used directly.
 *
 * @see com.facebook.FacebookActivity
 */
open class LoginFragment : Fragment() {
  private var callingPackage: String? = null
  private var request: LoginClient.Request? = null
  lateinit var loginClient: LoginClient
    private set
  lateinit var launcher: ActivityResultLauncher<Intent>
    private set

  private lateinit var progressBar: View

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    val restoredLoginClient = savedInstanceState?.getParcelable<LoginClient>(SAVED_LOGIN_CLIENT)
    loginClient =
        if (restoredLoginClient != null) {
          restoredLoginClient.fragment = this
          restoredLoginClient
        } else {
          createLoginClient()
        }
    loginClient.onCompletedListener =
        LoginClient.OnCompletedListener { outcome -> onLoginClientCompleted(outcome) }
    val activity = activity ?: return
    initializeCallingPackage(activity)
    val intent = activity.intent
    if (intent != null) {
      val bundle = intent.getBundleExtra(REQUEST_KEY)
      if (bundle != null) {
        request = bundle.getParcelable(EXTRA_REQUEST)
      }
    }

    launcher =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult(),
            getLoginMethodHandlerCallback(activity))
  }

  private fun getLoginMethodHandlerCallback(activity: FragmentActivity): (ActivityResult) -> Unit =
      { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
          loginClient.onActivityResult(
              LoginClient.getLoginRequestCode(), result.resultCode, result.data)
        } else {
          activity.finish()
        }
      }

  protected open fun createLoginClient(): LoginClient {
    return LoginClient(this)
  }

  override fun onDestroy() {
    loginClient.cancelCurrentHandler()
    super.onDestroy()
  }

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View? {
    val view = inflater.inflate(layoutResId, container, false)
    progressBar = view.findViewById<View>(R.id.com_facebook_login_fragment_progress_bar)
    loginClient.backgroundProcessingListener =
        object : LoginClient.BackgroundProcessingListener {
          override fun onBackgroundProcessingStarted() {
            showSpinner()
          }

          override fun onBackgroundProcessingStopped() {
            hideSpinner()
          }
        }
    return view
  }

  @get:LayoutRes
  protected open val layoutResId: Int
    get() = R.layout.com_facebook_login_fragment

  private fun onLoginClientCompleted(outcome: LoginClient.Result) {
    request = null
    val resultCode =
        if (outcome.code == LoginClient.Result.Code.CANCEL) Activity.RESULT_CANCELED
        else Activity.RESULT_OK
    val bundle = Bundle()
    bundle.putParcelable(RESULT_KEY, outcome)
    val resultIntent = Intent()
    resultIntent.putExtras(bundle)

    val hostActivity = activity
    // The activity might be detached we will send a cancel result in onDetach
    if (isAdded && hostActivity != null) {
      hostActivity.setResult(resultCode, resultIntent)
      hostActivity.finish()
    }
  }

  override fun onResume() {
    super.onResume()

    // If the calling package is null, this generally means that the callee was started
    // with a launchMode of singleInstance. Unfortunately, Android does not allow a result
    // to be set when the callee is a singleInstance, so we log an error and return.
    if (callingPackage == null) {
      Log.e(TAG, NULL_CALLING_PKG_ERROR_MSG)
      activity?.finish()
      return
    }
    loginClient.startOrContinueAuth(request)
  }

  override fun onPause() {
    super.onPause()
    val progressBar = view?.findViewById<View>(R.id.com_facebook_login_fragment_progress_bar)
    if (progressBar != null) {
      progressBar.visibility = View.GONE
    }
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    loginClient.onActivityResult(requestCode, resultCode, data)
  }

  override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    outState.putParcelable(SAVED_LOGIN_CLIENT, loginClient)
  }

  private fun showSpinner() {
    progressBar.visibility = View.VISIBLE
    onSpinnerShown()
  }

  private fun hideSpinner() {
    progressBar.visibility = View.GONE
    onSpinnerHidden()
  }

  protected open fun onSpinnerShown() {}

  protected open fun onSpinnerHidden() {}

  private fun initializeCallingPackage(activity: Activity) {
    val componentName = activity.callingActivity ?: return
    callingPackage = componentName.packageName
  }

  companion object {
    internal const val RESULT_KEY = "com.facebook.LoginFragment:Result"
    const val REQUEST_KEY = "com.facebook.LoginFragment:Request"
    const val EXTRA_REQUEST = "request"
    private const val TAG = "LoginFragment"
    private const val NULL_CALLING_PKG_ERROR_MSG =
        "Cannot call LoginFragment with a null calling package. This can occur if the launchMode of the caller is singleInstance."
    private const val SAVED_LOGIN_CLIENT = "loginClient"
  }
}
